EmailValidator/EmailLexer.php

```diff
@@ -122,8 +128,10 @@ public function getPrevious()
     public function moveNext()
     {
         $this->previous = $this->token;
+        $hasNext = parent::moveNext();
+        $this->token = $this->token ?: ['type' => null, 'value' => '', 'position' => 0];
```

Comment:
Is this line because of the lexer reaching the last token and being set to `null`?
I'd suggest to express that case with something like 
```php
public function moveNext()
{
    $this->previous = $this->token;
    $hasNext = parent::moveNext();
    /* new stuff */ $this->setEndToken($hasNext);
    return $hasNext;
}

private function setFinalToken($nextToken)
{
    $this->token = $this->token ?: ['type' => C_NUL, 'value' => '', 'position' => 0];
}
```

Reply:
you're right about the reason: $this->token cannot be null.
not sure I get what the additional method would provide thought :)

Reply:
To understood this line I had to review carefully the full PR and take my time, even with my knowledge of the library.
The additional method adds clarity on the reason behind.

Reply:
I pushed a variant, does it solve the concern?

Reply:
I think it can work.
Thanks!